### DISCUSSION TITLE
how SpringCloud Gateway add traceId to response header?

### DISCUSSION THREAD

**Discussion by ZhangZangQian** [apache/skywalking#10686] [Q&A] [answered]
how SpringCloud Gateway add traceId to response header?

**Comment by ZhangZangQian** [comment]
spring cloud gateway 3.1.x

**Comment by darknesstm** [comment]
```java
ClientHttpRequest clientHttpRequest = (ClientHttpRequest) allArguments[0];
        ContextCarrier contextCarrier = (ContextCarrier) objInst.getSkyWalkingDynamicField();
        CarrierItem next = contextCarrier.items();
        while (next.hasNext()) {
            next = next.next();
            clientHttpRequest.getHeaders().set(next.getHeadKey(), next.getHeadValue());
        }
```

[BodyInserterRequestInterceptor.java#L40](https://github.com/apache/skywalking-java/blob/02dad132b2807cbecdb69c7c0fbce54d36bd6e3e/apm-sniffer/apm-sdk-plugin/spring-plugins/spring-webflux-5.x-webclient-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/webflux/v5/webclient/BodyInserterRequestInterceptor.java#L40)

**Reply by ZhangZangQian** [reply]
Thank you for your answer, but I didn't understand it. Could you tell me in more detail, how should I code it in springcloud gateway?

**Reply by darknesstm** [reply]
Does this have anything to do with Skywalking?

You can just use the SCG Filter to add anything you want.

**Reply by lesliey** [reply]
@darknesstm I guess what  what you post is from java agent, but i need to get skywalking traceId in GatewayFilter and then put it into ServerWebExchange Resposne Headers.

**Reply by lesliey** [reply]
> @darknesstm I guess what what you post is from java agent, but i need to get skywalking traceId in GatewayFilter and then put it into ServerWebExchange Resposne Headers.

i mean that in SCG, i can not call ` objInst.getSkyWalkingDynamicField();` ; Will u post a code sample about a custom GatewayFilter?

**Reply by darknesstm** [reply, marked as answer]
`String traceId = WebFluxSkyWalkingOperators.continueTracing(exchange, TraceContext::traceId);`

exchange is `org.springframework.web.server.ServerWebExchange`

```xml
<dependency>
  <groupId>org.apache.skywalking</groupId>
  <artifactId>apm-toolkit-trace</artifactId>
  <version>${skywalking.version}</version>
</dependency>

<dependency>
  <groupId>org.apache.skywalking</groupId>
  <artifactId>apm-toolkit-webflux</artifactId>
  <version>${skywalking.version}</version>
</dependency>
```

**Reply by lesliey** [reply]
I have tried it, it shows `Could not find artifact org.apache.skywalking:apm-toolkit-webflux:pom:8.9.0 in central `。 U did upload it to maven central ?

**Reply by darknesstm** [reply]
upgrade to 8.15.0

**Comment by lesliey** [comment]
I just tried 8.15.0, but it does not work as also.

```
public class TraceFilter implements GlobalFilter, Ordered {

  @Override
  public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {

    String skyWalkingTraceId = WebFluxSkyWalkingOperators.continueTracing(exchange, TraceContext::traceId);
    String skyWalkingSegmentId = WebFluxSkyWalkingOperators.continueTracing(exchange, TraceContext::segmentId);
    int skyWalkingSpanId = WebFluxSkyWalkingOperators.continueTracing(exchange, TraceContext::spanId);
    log.info("SkyWalking traceId {} segmentId {} spanId {}", skyWalkingTraceId, skyWalkingSegmentId, skyWalkingSpanId);
  }

  @Override
  public int getOrder() {
    return Ordered.HIGHEST_PRECEDENCE;
  }

  }
```

prints as follow

```
2023-04-23 21:21:08.715  INFO 26572 --- [ctor-http-nio-3] TraceFilter  : SkyWalking traceId N/A segmentId N/A spanId -1
```

**Reply by darknesstm** [reply]
need `apm-toolkit-webflux-activation-8.15.0.jar` in agent's `activations` folder
maybe also need `apm-spring-webflux-5.x-plugin-8.15.0.jar` `apm-spring-cloud-gateway-3.x-plugin-8.15.0.jar` in agent's `plugins` folder

**Reply by lesliey** [reply]
Yeap, it works well finally! Thanks a lot. Maybe i should post a blog to share it.

**Comment by collery** [comment]
hello , if don't  use  WebFluxSkyWalkingOperators.continueTracing  and not import webflux plugins . what will happen?

**Reply by collery** [reply]
@darknesstm     we only use  exchange.getResponse().getHeaders().set("X-Trace-ID", TraceContext.traceId())   then found  tid sometime loss 😄

**Reply by darknesstm** [reply]
`TraceContext.traceId()` is based on a `ThreadLocal`, and if there is no appropriate context copying across threads, the data will be lost.

**Reply by collery** [reply]
> `TraceContext.traceId()`是基于a的`ThreadLocal`，如果没有合适的context跨线程copy，数据就会丢失。

thanks
